From 58e53cb059a86cd9053cffe7f65fc9e699a1fd48 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Mon, 10 Feb 2025 05:03:11 -0500
Subject: [PATCH 793/954] feat(stm32mp1): move DDR training area restore on
 low-power resume

Restore DDR training area on resume after encryption was enabled,
so execute this step at platform level.
On STM32MP13, if DDRMCE region starts on its base address, the DDR
training area is included in this region. It must be restored in the
same conditions as when it was saved, so with the encryption enabled.
Keep a generic sequence for all STM32MP1 products.
Encapsulate the restore memcpy with DCACHE invalidate and flush to
ensure restore completion.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git c7968e3558478a064b3567b7c22ab5f4fde8a1cd feat(stm32mp1): move DDR training area restore on low-power resume]
Co-authored-by: Dirk Schoebel <dirk@liji.de>
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: Ic07352f4572673ed806a91b641b27728e093b21b
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/435911
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp1/bl2_plat_setup.c   | 8 +++++++-
 plat/st/stm32mp1/stm32mp1_context.c | 7 ++++++-
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/plat/st/stm32mp1/bl2_plat_setup.c b/plat/st/stm32mp1/bl2_plat_setup.c
index 988397bc47..6d231d77a4 100644
--- a/plat/st/stm32mp1/bl2_plat_setup.c
+++ b/plat/st/stm32mp1/bl2_plat_setup.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015-2023, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2015-2025, Arm Limited and Contributors. All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -589,6 +589,12 @@ int bl2_plat_handle_post_image_load(unsigned int image_id)
 			prepare_encryption();
 		}
 #endif
+
+		/* Restore DDR training area on resume after encryption was enabled */
+		if (wakeup_ddr_sr) {
+			stm32_restore_ddr_training_area();
+		}
+
 		if (stm32mp_skip_boot_device_after_standby()) {
 			return 0;
 		}
diff --git a/plat/st/stm32mp1/stm32mp1_context.c b/plat/st/stm32mp1/stm32mp1_context.c
index 877c435cf4..1c89703cbc 100644
--- a/plat/st/stm32mp1/stm32mp1_context.c
+++ b/plat/st/stm32mp1/stm32mp1_context.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017-2023, ARM Limited and Contributors. All rights reserved.
+ * Copyright (c) 2017-2025, ARM Limited and Contributors. All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -188,9 +188,14 @@ void stm32_restore_ddr_training_area(void)
 
 	backup_data = (struct backup_data_s *)STM32MP_BACKUP_RAM_BASE;
 
+	inv_dcache_range(STM32MP_DDR_BASE, TRAINING_AREA_SIZE);
+
 	memcpy((uint32_t *)STM32MP_DDR_BASE,
 	       &backup_data->ddr_training_backup,
 	       TRAINING_AREA_SIZE);
+
+	flush_dcache_range(STM32MP_DDR_BASE, TRAINING_AREA_SIZE);
+
 	dsb();
 
 	clk_disable(BKPSRAM);
-- 
2.53.0

